### PR TITLE
Compares versions in place of template folder numbers

### PR DESCRIPTION
**Pull request by mrajashree**
upgradeVersionLinks is updated by comparing versions in place of template's folder numbers.

### CODE REVIEW

File: manager/catalog_manager.go [hunk 1]
```diff
@@ -328,17 +335,40 @@ func ReadTemplateVersion(catalogID string, templateID string, versionID string)
 func GetNewTemplateVersions(path string) (model.Template, bool) {
 	templateMetadata := model.Template{}
 
+	log.Debugf("Over here path : %v\n", path)
```
Comment (vincent99) [inline manager/catalog_manager.go]: Remove debugging logs statements, or make the messages something that will be meaningful if you read it in a log a month from now.

File: manager/catalog_manager.go [hunk 2]
```diff
@@ -328,17 +335,40 @@ func ReadTemplateVersion(catalogID string, templateID string, versionID string)
...
+	rancherComposePathCurrent := CatalogRootDir + catalogID + "/templates/" + parentPath + "/" + cVersion
+	log.Debugf("Exact path : %s\n", rancherComposePathCurrent)
+
+	composeBytes, err := readFile(rancherComposePathCurrent, "rancher-compose.yml")
+	if err != nil {
+		return templateMetadata, false
+	}
+
+	//read the questions section
+	RC := make(map[string]model.RancherCompose)
```
Comment (vincent99) [inline manager/catalog_manager.go]: local variables should start with lowercase (`rc`)

File: manager/catalog_manager.go [hunk 3]
```diff
@@ -328,17 +335,40 @@ func ReadTemplateVersion(catalogID string, templateID string, versionID string)
...
+	err = yaml.Unmarshal(*composeBytes, &RC)
+	if err != nil {
+		log.Errorf("Error unmarshalling %s under template: %s, error: %v", "rancher-compose.yml", rancherComposePathCurrent, err)
+		return templateMetadata, false
+	}
+	current_version := RC[".catalog"].Version
```
Comment (vincent99) [inline manager/catalog_manager.go]: and be camelCased (`currentVersion`)

File: manager/catalog_manager.go [hunk 4]
```diff
@@ -328,17 +335,40 @@ func ReadTemplateVersion(catalogID string, templateID string, versionID string)
...
+	if strings.Count(current_version, ".") < 2 {
+		current_version += ".0"
+	}
+	vCurr, err := getSemVersion(current_version)
 
-	currentVersion, err := strconv.Atoi(cVersion)
+	// currentVersion, err := strconv.Atoi(cVersion)
```
Comment (vincent99) [inline manager/catalog_manager.go]: Remove obsolete code, it's still in git if you ever need it.

File: manager/catalog_manager.go [hunk 5]
```diff
@@ -328,17 +335,40 @@ func ReadTemplateVersion(catalogID string, templateID string, versionID string)
...
+
+	//read the questions section
+	RC := make(map[string]model.RancherCompose)
+	err = yaml.Unmarshal(*composeBytes, &RC)
+	if err != nil {
+		log.Errorf("Error unmarshalling %s under template: %s, error: %v", "rancher-compose.yml", rancherComposePathCurrent, err)
+		return templateMetadata, false
+	}
+	current_version := RC[".catalog"].Version
+	if strings.Count(current_version, ".") < 2 {
```
Comment (vincent99) [inline manager/catalog_manager.go]: This should go in getSemVersion so it's not repeated.

Needs to be a little more robust to work for an input like `v1.2-pre1` -> `v1.2.0-pre1`, not `v1.2-pre1.0`

File: manager/catalog_manager.go [hunk 6]
```diff
@@ -347,13 +377,33 @@ func GetNewTemplateVersions(path string) (model.Template, bool) {
...
 			copyOfversionLinks := make(map[string]string)
+
 			for key, value := range templateMetadata.VersionLinks {
 				if value != path {
 					otherVersionTokens := strings.Split(value, ":")
 					oVersion := otherVersionTokens[2]
-					otherVersion, err := strconv.Atoi(oVersion)
 
-					if err == nil && otherVersion > currentVersion {
+					rancherComposePathOther := CatalogRootDir + catalogID + "/templates/" + parentPath + "/" + oVersion
```
Comment (vincent99) [inline manager/catalog_manager.go]: All this is repeated too, combine into a single function something like `getVersionForPath(path string) string {}`that reads a file, unmarshalls, and returns semversion.

File: manager/catalog_manager.go [hunk 7]
```diff
@@ -368,3 +418,15 @@ func GetNewTemplateVersions(path string) (model.Template, bool) {
 	log.Debugf("Template metadata not found by path: %s", path)
 	return templateMetadata, false
 }
+
+func getSemVersion(versionStr string) (*semver.Version, error) {
+	log.Debugf("Before re : %s\n", versionStr)
+	versionStr = re.ReplaceAllString(versionStr, "$1")
```
Comment (vincent99) [inline manager/catalog_manager.go]: Regex unnecessary for removing one character..

```
if strings.Index(versionStr,"v") == 0 {
  versionStr = versionStr[1:]
}
```

File: manager/catalog_manager.go [hunk 8]
```diff
@@ -335,10 +336,17 @@ func GetNewTemplateVersions(path string) (model.Template, bool) {
...
+	rancherComposePathCurrent := CatalogRootDir + catalogID + "/templates/" + parentPath + "/" + cVersion
 
-	currentVersion, err := strconv.Atoi(cVersion)
+	currentVersion, err := getVersionFromRancherCompose(rancherComposePathCurrent)
+	if err != nil {
+		log.Errorf("Error %v getting semVersion ", err)
+		return templateMetadata, false
+	}
 
 	if err != nil {
```
Comment (vincent99) [inline manager/catalog_manager.go]: This is unreachable, if err != nil you've already returned above.

File: manager/catalog_manager.go [hunk 9]
```diff
@@ -368,3 +383,35 @@ func GetNewTemplateVersions(path string) (model.Template, bool) {
...
+
+	err = yaml.Unmarshal(*composeBytes, &rancherCompose)
+	if err != nil {
+		log.Errorf("Error unmarshalling %s under template: %s, error: %v", "rancher-compose.yml", rancherComposePath, err)
+		return nil, err
+	}
+	version := rancherCompose[".catalog"].Version
+	if strings.Count(version, ".") < 2 {
+		preReleaseIndex := strings.Index(version, "-")
+		version = version[:preReleaseIndex] + ".0" + version[preReleaseIndex:]
```
Comment (vincent99) [inline manager/catalog_manager.go]: This will panic on a string that has 0-1 `.` and no`-`, e.g. `v1.0`, `v1`

**Comment by mrajashree** [comment]
@vincent99 Did those changes, can you review it again?

**Comment by vincent99** [comment]
One more thing, there should be a test that exercises the semver parsing (that would have caught the panic on `v1.0`) and ensures it returns the all of (and only) the expected list for each version.

It is very important that this version filtering works properly to limit installs to the correct upgrades... Offering an "upgrade" from e.g. Kubernetes 1.3.x to 1.2.y will break someone if they click it and don't read carefully.

You can do something like hand-create a $sorted list of all the versions in your qa-catalog template, get the template from the API and loop over all the templateVersions.   For each version:
 - Find the $index in the $sorted array for this version
 - Assert that there are the expected number of entries in the upgradeVersionLinks map.. length(keys of upgradeVersionLinks)) ==  length($sorted) - $index -1
 - And that each upgradeVersionLinks key returned has an index in the $sorted array >= current version $index

**Comment by mrajashree** [comment]
@vincent99 Added the test

**Comment by mrajashree** [comment]
https://github.com/rancher/rancher/issues/5691

**Pull request merged**